soundbay/active_learning.py

```diff
@@ -0,0 +1,165 @@
+    """
+    chunk_bool_mask, chunk_end_sec = compute_chunk_boundaries(chunk_len_in_seconds, chunk_start_sec, df, recording_id,
+                                                              recording_len_in_secs)
+    chunk_id = f'recording_{recording_id}_sec_{chunk_start_sec}_to_{chunk_end_sec}'
+    df.loc[chunk_bool_mask, 'chunk_id'] = chunk_id
+    return df
+
+
+def compute_chunk_boundaries(chunk_len_in_seconds: int, chunk_start_sec: int, df: pd.DataFrame, recording_id: str,
+                             recording_len_in_secs: int):
+    """
+    Figure out which segments should be included in chunk.
+    :return:
+        chunk_bool_mask: a boolean mask of all segments (=rows) included in chunk; pd.Series
+        chunk_end_sec: the last recording second included in the chunk; int
+    """
+    chunk_end_sec = min(chunk_start_sec + chunk_len_in_seconds, recording_len_in_secs)
+    chunk_bool_mask = ((df.recording == recording_id) & (df.segment_start_sec >= chunk_start_sec) & (
+            df.segment_end_sec <= chunk_end_sec))
+    chunk_actual_size = df[chunk_bool_mask].shape[0]
+    chunk_end_sec = validate_chunk_end_sec(chunk_end_sec, chunk_start_sec, chunk_actual_size)
+    return chunk_bool_mask, chunk_end_sec
+
+
+def compute_potential_gain_per_segment(df: pd.DataFrame) -> pd.DataFrame:
+    df['max'] = df[['class0_prob', 'class1_prob']].max(axis=1)
+    df['gain'] = 1 - df['max']
+    df = df.drop('max', axis=1)
+    return df
+
+
+def n_high_priority_segments(df_chunk: pd.DataFrame, threshold) -> int:
+    """
+    returns the number of high priority segments (where gain > threshold) in given dataframe
+    """
+    return df_chunk[df_chunk['gain'] > threshold].shape[0]
+
+
+def format_ranked_chunks_for_output(ranked_chunks: pd.Series) -> pd.DataFrame:
+    '''
+    Reformat a Pandas series of ranked chunks into the dataframe that serves as final output.
+    :param ranked_chunks: pandas series containing chunk_id (as index) and the chunk gain score (the higher score, the more valuable the chunk).
+    :return: pandas DataFrame with chunk_id (as index), chunk score (see above), and ranking (starting at 1; the lower the ranking, the more valuable the chunk).
+    '''
+    df_ranked_chunks = pd.DataFrame(ranked_chunks)
+    df_ranked_chunks.columns = ['chunk_score']
+    df_ranked_chunks.insert(0, 'ranking', np.arange(df_ranked_chunks.shape[0]) + 1)
+    return df_ranked_chunks
+
+
+def get_ranked_segments_from_inference_dir(inference_dir: str, chunk_len_in_seconds: int,
+                                           gain_threshold: float) -> pd.DataFrame:
+    df = load_inference_results_from_dir(inference_dir)
+    for recording_id in df.recording.unique():
+        df = add_chunk_id_per_recording(df, recording_id, chunk_len_in_seconds)
+    df = compute_potential_gain_per_segment(df)
+    ranked_chunks = df.groupby('chunk_id').apply(
+        lambda x: n_high_priority_segments(x, threshold=gain_threshold)).sort_values(
```

Comment:
Indeed, why use a threshold and not n (n being "number of segments which we are willing to annotate" or such).